// ionic2-auto-complete/ionic2-auto-complete.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('rxjs'), require('ionic-angular')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@angular/forms', 'rxjs', 'ionic-angular'], factory) :
	(factory((global['ionic2-auto-complete'] = global['ionic2-auto-complete'] || {}),global._angular_core,global._angular_common,global._angular_forms,global.rxjs,global.ionicAngular));
}(this, (function (exports,_angular_core,_angular_common,_angular_forms,rxjs,ionicAngular) { 'use strict';

// searchbar default options
var defaultOpts = {
    cancelButtonText: 'Cancel',
    showCancelButton: false,
    debounce: 250,
    placeholder: 'Search',
    autocomplete: 'off',
    autocorrect: 'off',
    spellcheck: 'off',
    type: 'search',
    value: '',
    noItems: '',
    clearOnEdit: false,
    clearInput: false
};
var AutoCompleteComponent = (function () {
    /**
     * create a new instace
     */
    function AutoCompleteComponent() {
        this.keyword = null;
        this.suggestions = [];
        this.showList = false;
        this.itemSelected = new _angular_core.EventEmitter();
        this.ionAutoInput = new _angular_core.EventEmitter();
        this.options = {};
        // set default options
        this.defaultOpts = defaultOpts;
    }
    /**
     * get items for auto-complete
     * @return {?}
     */
    AutoCompleteComponent.prototype.getItems = function () {
        var _this = this;
        if (this.showResultsFirst && !this.keyword) {
            this.keyword = '';
        }
        else if (this.keyword.trim() === '') {
            this.suggestions = [];
            return;
        }
        var /** @type {?} */ result = this.dataProvider.getResults(this.keyword);
        // if query is async
        if (result instanceof rxjs.Observable) {
            result
                .subscribe(function (results) {
                _this.suggestions = results;
                _this.showItemList();
            }, function (error) { return console.error(error); });
        }
        else {
            this.suggestions = result;
            this.showItemList();
        }
        // emit event
        this.ionAutoInput.emit(this.keyword);
    };
    /**
     * show item list
     * @return {?}
     */
    AutoCompleteComponent.prototype.showItemList = function () {
        this.showList = true;
    };
    /**
     * hide item list
     * @return {?}
     */
    AutoCompleteComponent.prototype.hideItemList = function () {
        this.showList = false;
    };
    /**
     * select item from list
     * @param {?} selection
     * @return {?}
     */
    AutoCompleteComponent.prototype.select = function (selection) {
        this.keyword = this.dataProvider.labelAttribute == null || selection[this.dataProvider.labelAttribute] == null
            ? selection : selection[this.dataProvider.labelAttribute];
        this.hideItemList();
        // emit selection event
        this.itemSelected.emit(selection);
    };
    /**
     * get current input value
     * @return {?}
     */
    AutoCompleteComponent.prototype.getValue = function () {
        return this.keyword;
    };
    /**
     * set current input value
     * @param {?} value
     * @return {?}
     */
    AutoCompleteComponent.prototype.setValue = function (value) {
        this.keyword = value;
        return;
    };
    /**
     * /**
     * clear current input value
     * @param {?=} hideItemList
     * @return {?}
     */
    AutoCompleteComponent.prototype.clearValue = function (hideItemList) {
        if (hideItemList === void 0) { hideItemList = false; }
        this.keyword = null;
        if (hideItemList) {
            this.hideItemList();
        }
        return;
    };
    /**
     * handle document click
     * @param {?} event
     * @return {?}
     */
    AutoCompleteComponent.prototype.documentClickHandler = function (event) {
        var /** @type {?} */ target = event.target;
        var /** @type {?} */ parent = event.target.parentElement;
        if (target.className.split(" ").indexOf("ion-auto-complete") == -1 && parent.className.split(" ").indexOf("ion-auto-complete") == -1) {
            this.hideItemList();
        }
    };
    return AutoCompleteComponent;
}());
AutoCompleteComponent.decorators = [
    { type: _angular_core.Component, args: [{
                host: {
                    '(document:click)': 'documentClickHandler($event)',
                },
                template: "\n        <ion-input (keyup)=\"getItems($event)\" *ngIf=\"useIonInput\"\n                     (tap)=\"showResultsFirst && getItems()\"\n                     [(ngModel)]=\"keyword\"\n                     [placeholder]=\"options.placeholder == null ? defaultOpts.placeholder : options.placeholder\"\n                     [type]=\"options.type == null ? defaultOpts.type : options.type\"\n                     [clearOnEdit]=\"options.clearOnEdit == null ? defaultOpts.clearOnEdit : options.clearOnEdit\"\n                     [clearInput]=\"options.clearInput == null ? defaultOpts.clearInput : options.clearInput\"\n                     [ngClass]=\"['ion-auto-complete']\"\n                     >\n      </ion-input>\n      <ion-searchbar (ionInput)=\"getItems($event)\" *ngIf=\"!useIonInput\"\n                     (tap)=\"showResultsFirst && getItems()\"\n                     [(ngModel)]=\"keyword\"\n                     [cancelButtonText]=\"options.cancelButtonText == null ? defaultOpts.cancelButtonText : options.cancelButtonText\"\n                     [showCancelButton]=\"options.showCancelButton == null ? defaultOpts.showCancelButton : options.showCancelButton\"\n                     [debounce]=\"options.debounce == null ? defaultOpts.debounce : options.debounce\"\n                     [placeholder]=\"options.placeholder == null ? defaultOpts.placeholder : options.placeholder\"\n                     [autocomplete]=\"options.autocomplete == null ? defaultOpts.autocomplete : options.autocomplete\"\n                     [autocorrect]=\"options.autocorrect == null ? defaultOpts.autocorrect : options.autocorrect\"\n                     [spellcheck]=\"options.spellcheck == null ? defaultOpts.spellcheck : options.spellcheck\"\n                     [type]=\"options.type == null ? defaultOpts.type : options.type\"\n                     [ngClass]=\"['ion-auto-complete']\"\n      >\n      </ion-searchbar>\n      <ng-template #defaultTemplate let-attrs=\"attrs\">\n          <span [innerHTML]='(attrs.labelAttribute ? attrs.data[attrs.labelAttribute] : attrs.data) | boldprefix:attrs.keyword'></span>\n      </ng-template>\n      <ul *ngIf=\"suggestions.length > 0 && showList\">\n          <li *ngFor=\"let suggestion of suggestions\" (tap)=\"select(suggestion)\">\n              <ng-template \n                        [ngTemplateOutlet]=\"template || defaultTemplate\" \n                        [ngOutletContext]=\"\n                        {attrs:{ data: suggestion, keyword: keyword, labelAttribute: dataProvider.labelAttribute }}\"></ng-template>\n          </li>\n      </ul>\n      <p *ngIf=\"suggestions.length == 0 && showList && options.noItems\">{{ options.noItems }}</p>\n  ",
                selector: 'ion-auto-complete'
            },] },
];
/**
 * @nocollapse
 */
AutoCompleteComponent.ctorParameters = function () { return []; };
AutoCompleteComponent.propDecorators = {
    'dataProvider': [{ type: _angular_core.Input },],
    'options': [{ type: _angular_core.Input },],
    'keyword': [{ type: _angular_core.Input },],
    'showResultsFirst': [{ type: _angular_core.Input },],
    'template': [{ type: _angular_core.Input },],
    'useIonInput': [{ type: _angular_core.Input },],
    'itemSelected': [{ type: _angular_core.Output },],
    'ionAutoInput': [{ type: _angular_core.Output },],
};

/**
 * bolds the beggining of the matching string in the item
 */
var BoldPrefix = (function () {
    function BoldPrefix() {
    }
    /**
     * @param {?} value
     * @param {?} keyword
     * @return {?}
     */
    BoldPrefix.prototype.transform = function (value, keyword) {
        return value.replace(new RegExp(keyword, 'job'), function (str) {
             return str.bold(); }
             );
    };
    return BoldPrefix;
}());
BoldPrefix.decorators = [
    { type: _angular_core.Pipe, args: [{ 
                name: 'boldprefix'
            },] },
    { type: _angular_core.Injectable },
];
/**
 * @nocollapse
 */
BoldPrefix.ctorParameters = function () { return []; };

var AutoCompleteModule = (function () {
    function AutoCompleteModule() {
    }
    /**
     * @return {?}
     */
    AutoCompleteModule.forRoot = function () {
        return {
            ngModule: AutoCompleteModule,
            providers: []
        };
    };
    return AutoCompleteModule;
}());
AutoCompleteModule.decorators = [
    { type: _angular_core.NgModule, args: [{
                imports: [
                    _angular_common.CommonModule,
                    _angular_forms.FormsModule,
                    ionicAngular.IonicModule
                ],
                declarations: [
                    AutoCompleteComponent,
                    BoldPrefix
                ],
                exports: [
                    AutoCompleteComponent,
                    BoldPrefix
                ]
            },] },
];
/**
 * @nocollapse
 */
AutoCompleteModule.ctorParameters = function () { return []; };

exports.AutoCompleteModule = AutoCompleteModule;
exports.AutoCompleteComponent = AutoCompleteComponent;
exports.BoldPrefix = BoldPrefix;

Object.defineProperty(exports, '__esModule', { value: true });

})));
